pkg/controller/ttlafterfinished/ttlafterfinished_controller.go

```diff
@@ -0,0 +1,441 @@
+	recorder record.EventRecorder
+
+	// jLister can list/get Jobs from the shared informer's store
+	jLister batchlisters.JobLister
+
+	// jStoreSynced returns true if the Job store has been synced at least once.
+	// Added as a member to the struct to allow injection for testing.
+	jListerSynced cache.InformerSynced
+
+	// Jobs that the controller will attempt to delete.
+	jobsToCleanup workqueue.RateLimitingInterface
+
+	// Jobs to be added to job cleanup workqueue later after it's ready for clean up.
+	// These jobs are stored in a thread-safe heap.
+	jobsDelayCleanupQueueMutex sync.RWMutex
+	jobsDelayCleanupQueue      *cache.Heap
+
+	// The next time we should check the jobsDelayCleanupQueue.
+	nextCheckAtMutex sync.RWMutex
+	nextCheckAt      *time.Time
+
+	// The clock for tracking time
+	clock clock.Clock
+}
+
+// New creates an instance of TTLAfterFinishedController
+func New(jobInformer batchinformers.JobInformer, client clientset.Interface) *TTLAfterFinishedController {
+	eventBroadcaster := record.NewBroadcaster()
+	eventBroadcaster.StartLogging(glog.Infof)
+	eventBroadcaster.StartRecordingToSink(&v1core.EventSinkImpl{Interface: client.CoreV1().Events("")})
+
+	if client != nil && client.CoreV1().RESTClient().GetRateLimiter() != nil {
+		metrics.RegisterMetricAndTrackRateLimiterUsage("ttl_after_finished_controller", client.CoreV1().RESTClient().GetRateLimiter())
+	}
+
+	tc := &TTLAfterFinishedController{
+		client:        client,
+		recorder:      eventBroadcaster.NewRecorder(scheme.Scheme, v1.EventSource{Component: "ttl-after-finished-controller"}),
+		jobsToCleanup: workqueue.NewNamedRateLimitingQueue(workqueue.DefaultControllerRateLimiter(), "ttl_jobs_to_delete"),
+	}
+
+	jobInformer.Informer().AddEventHandler(cache.ResourceEventHandlerFuncs{
+		AddFunc:    tc.addJob,
+		UpdateFunc: tc.updateJob,
+		DeleteFunc: tc.deleteJob,
+	})
+
+	tc.jLister = jobInformer.Lister()
+	tc.jListerSynced = jobInformer.Informer().HasSynced
+
+	tc.jobsDelayCleanupQueue = NewJobsDelayCleanupQueue()
+
+	tc.clock = clock.RealClock{}
+
+	return tc
+}
+
+// Run starts the workers to clean up Jobs.
+func (tc *TTLAfterFinishedController) Run(workers int, stopCh <-chan struct{}) {
+	defer utilruntime.HandleCrash()
+	defer tc.jobsToCleanup.ShutDown()
+	defer tc.jobsDelayCleanupQueue.Close()
+
+	glog.Infof("Starting TTL after finished controller")
+	defer glog.Infof("Shutting down TTL after finished controller")
+
+	if !controller.WaitForCacheSync("TTL after finished", stopCh, tc.jListerSynced) {
+		return
+	}
+
+	for i := 0; i < workers; i++ {
+		go wait.Until(tc.runJobCleanupWorker, time.Second, stopCh)
+	}
+	go wait.JitterUntil(tc.runJobDelayCleanupWorker, 10*time.Second, 0.2, true, stopCh)
+
+	<-stopCh
+}
+
+func (tc *TTLAfterFinishedController) addJob(obj interface{}) {
+	job := obj.(*batch.Job)
+	glog.V(4).Infof("Adding job %s", job.Name)
+
+	// We don't care about the Jobs that's going to be deleted.
+	if job.DeletionTimestamp != nil {
+		return
+	}
+
```

Comment:
Shouldn't we just directly filter unmanaged jobs here with  `TTLSecondsAfterFinished == nil`? Same goes for the other functions.

Reply:
It's done in `checkTTLAndReact`. Actually, it makes sense to move the DeletionTimestamp check there too. Updated comments to make it more clear.